src/shared/install.c

```diff
@@ -2868,15 +2868,70 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
         return 0;
 }
 
-static int query_presets(const char *name, const Presets presets) {
+static int pattern_match_multiple_instances(
+                        const char *pattern,
+                        const char *unit_name,
+                        char ***preset_array) {
+
+        const char *parameter;
+        char **iter;
+        _cleanup_free_ char *templated_name = NULL, *prefix = NULL, *instance_name = NULL;
+        int ret;
+
+        if (!preset_array || !unit_name_is_valid(unit_name, UNIT_NAME_TEMPLATE))
+                return -1;
+
+        /* We check the first word of preset pattern to see if it matches
+         * the template name */
+        ret = unit_name_template(unit_name, &templated_name);
+        if (ret < 0)
+                return ret;
+        parameter = first_word(pattern, templated_name);
+
+        if (!parameter)
+                return -1;
+
+        _cleanup_strv_free_ char **l = NULL;
+        l = strv_split(parameter, WHITESPACE);
+
+        /* We want the instance part of the service, and see if it matches in the list */
+        ret = unit_name_to_instance(unit_name, &instance_name);
+        if (ret < 0)
+                return ret;
+
+        /* If the unit name is found in the specified multiple instances, return matching with a null list */
+        if(strv_find(l, instance_name))
+                return 0;
+
+        /* Compose a list of specified instances when instance_name of the unit is empty  */
+        if (isempty(instance_name)) {
+                ret = unit_name_to_prefix(unit_name, &prefix);
+
+                _cleanup_strv_free_ char **out_strv = NULL;
+                STRV_FOREACH(iter, l) {
+                        _cleanup_free_ char *test;
```

Comment:
This needs to initalized `= NULL`, because otherwise the uninitialized pointer may be freed on error.